File: 5. Validations and transformations for backend engineers/code/typescript/example_1.ts
import { z } from 'zod';
import { Request, Response } from 'express';

// CreateBook is the SCHEMA. Each annotation = one rule
const CreateBookSchema = z.object({
  name: z.string().min(5).max(100)
});
type CreateBook = z.infer<typeof CreateBookSchema>;

// runPipeline decodes then validates; returns 400-ready messages.
export function runPipeline(req: Request, res: Response): CreateBook | null {
  try {
    // constructing the model runs all three layers
    // in order: existence -> type -> constraint
    return CreateBookSchema.parse(req.body);
  } catch (e) {
    if (e instanceof z.ZodError) {
      const messages = e.errors.map(err => `${err.path.join('.')}: ${err.message}`);
      res.status(400).json({ detail: messages });
      return null;
    }
    throw e;
  }
}

File: 5. Validations and transformations for backend engineers/code/rust/example_1.rs
use actix_web::{web, HttpResponse, ResponseError};
use serde::{Deserialize, Serialize};
use validator::Validate;
use std::fmt;

// CreateBook is the SCHEMA. Each annotation = one rule:
#[derive(Debug, Deserialize, Validate)]
pub struct CreateBook {
    #[validate(length(min = 5, max = 100))]
    pub name: String,
}

// runPipeline decodes then validates; returns 400-ready messages.
pub async fn run_pipeline(body: web::Json<CreateBook>) -> Result<CreateBook, actix_web::Error> {
    let create_book = body.into_inner();
    
    // 2. EXISTENCE + CONSTRAINT checks in one pass.
    if let Err(e) = create_book.validate() {
        // reshape errors into clean strings
        return Err(actix_web::error::ErrorBadRequest(e.to_string()));
    }
    Ok(create_book)
}
